theseus/embodied/kinematics/kinematics_model.py

```diff
@@ -0,0 +1,112 @@
+
+
+class KinematicsModel(abc.ABC):
+    def __init__(self):
+        pass
+
+    @abc.abstractmethod
+    def forward_kinematics(self, robot_pose: RobotModelInput) -> Dict[str, LieGroup]:
+        pass
+
+    @abc.abstractmethod
+    def dim(self) -> int:
+        pass
+
+
+class IdentityModel(KinematicsModel):
+    def __init__(self):
+        super().__init__()
+
+    def forward_kinematics(self, robot_pose: RobotModelInput) -> Dict[str, LieGroup]:
+        if isinstance(robot_pose, Point2) or isinstance(robot_pose, Vector):
+            assert robot_pose.dof() == 2
+            return {"state": robot_pose}
+        raise NotImplementedError(
+            f"IdentityModel not implemented for pose with type {type(robot_pose)}."
+        )
+
+    def dim(self) -> int:
+        return 1
+
+
+class UrdfRobotModel(KinematicsModel):
+    def __init__(self, urdf_path: str):
+        self.drm_model = drm.DifferentiableRobotModel(urdf_path)
+
+    def _postprocess_quaternion(self, quat):
+        # Convert quaternion convention (DRM uses xyzw, Theseus uses wxyz)
+        quat_converted = torch.cat([quat[..., 3:], quat[..., :3]], dim=-1)
+
+        # Normalize quaternions
+        quat_normalized = quat_converted / torch.linalg.norm(
+            quat_converted, dim=-1, keepdim=True
+        )
+
+        return quat_normalized
+
+    def forward_kinematics(
+        self,
+        joint_states: RobotModelInput,
+        jacobians: Dict[str, Optional[torch.Tensor]] = {},
+    ) -> Dict[str, LieGroup]:
+        """Computes forward kinematics
+        Args:
+            joint_states: Vector of all joint angles
+        Outputs:
+            Dictionary that maps link name to link pose
+        """
+        # Check input dimensions
+        robot_model_dofs = len(self.drm_model.get_joint_limits())
+        assert joint_states.shape[-1] == robot_model_dofs, (
+            f"Robot model dofs ({robot_model_dofs}) incompatible with "
+            f"input joint state dimensions ({joint_states.shape[-1]})."
+        )
+
+        # Parse input
+        if isinstance(joint_states, torch.Tensor):
+            joint_states_input = joint_states
+        elif isinstance(joint_states, Variable):
+            joint_states_input = joint_states.data
+        else:
+            raise Exception(
+                "Invalid input joint states data type. Valid types are torch.Tensor and th.Vector."
+            )
+
+        # Compute forward kinematics for all links
+        link_poses: Dict[str, LieGroup] = {}
+        for link_name in self.drm_model.get_link_names():
+            pos, quat = self.drm_model.compute_forward_kinematics(
+                joint_states_input, link_name
+            )
+            quat_processed = self._postprocess_quaternion(quat)
+
+            link_poses[link_name] = SE3(
+                x_y_z_quaternion=torch.cat([pos, quat_processed], dim=-1)
+            )
+
+        # Compute jacobians
+        for link_name in jacobians.keys():
```

Comment:
We should add a `None` check and skip computing jacobians when not needed to be more efficient.

Reply:
For consistency we could add an identity jacobian to be returned for the `IdentityModel` and add the jacobians input signature to the base class.

Reply:
The idea is to only pass in a dictionary containing only keys (link_names) whose jacobians are desired.

Reply:
I see then this is implicitly handled by passing only the required link names.